types/ember__component/test/helper.ts

```diff
@@ -21,12 +21,80 @@ class Timestamp extends Helper<{
     }
 }
 
-// $ExpectType FunctionBasedHelper<{ PositionalArgs: [number, number]; NamedArgs: Record<string, unknown>; Return: number; }>
-const addHelper = helper(function add([a, b]: [number, number]) {
+interface DemoSig {
+    Args: {
+        Named: {
+            name: string;
+            age: number;
+        };
+        Positional: [i18nizer: (s: string) => string];
+    };
+    Return: string;
+}
+
+class SignatureForm extends Helper<DemoSig> {
+    compute([i18nizer]: [i18nizer: (s: string) => string], { name, age }: { name: string; age: number }): string {
+        return i18nizer(`${name} is ${age} years old`);
+    }
+}
+
+class BadPosSigForm extends Helper<DemoSig> {
+    compute(
+        // $ExpectError
+        [i18nizer, extra]: [i18nizer: (s: string) => string],
+        { name, age }: { name: string; age: number },
+    ): string {
+        return i18nizer(`${name} is ${age} years old`);
+    }
+}
+
+class BadNamedSigForm extends Helper<DemoSig> {
+    compute(
+        [i18nizer]: [i18nizer: (s: string) => string],
+        // $ExpectError
+        { name, age, potato }: { name: string; age: number },
+    ): string {
+        return i18nizer(`${name} is ${age} years old`);
+    }
+}
+
+class BadReturnForm extends Helper<DemoSig> {
+    compute([i18nizer]: [i18nizer: (s: string) => string], { name, age }: { name: string; age: number }): string {
+        // $ExpectError
+        return Boolean(i18nizer(`${name} is ${age} years old`));
+    }
+}
+
+// $ExpectType FunctionBasedHelper<{ Args: { Positional: [number, number]; Named: DefaultNamed; }; Return: number; }>
```

Comment:
Ah, that makes sense. We need to make sure, then, that we *don’t* use `DefaultNamed` anywhere *except* in that position, I think?

Reply:
I think it's ok anywhere we have `extends DefaultNamed` currently as well, since nearly _any_ object type will meet that requirement.

That said, the type param for the positional params is the only one I'm convinced really needs a constraint. I'd happily see the others go away 😄

Reply:
Wellllll getting it working just right involved shenanigans I do not like, but I *did* get it working correctly. (Further thoughts welcome, of course!) The shenanigans had to do with inference for helpers, *not* with your suggestion: TS reasonably wants to default an unspecified parameter in a function to `unknown`, so we need to actually provide a default of `N = EmptyObject` for the inference-driven case.